### DISCUSSION TITLE
Where should database failures be handled?

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1370] [General]
See https://github.com/fedimint/fedimint/pull/348#discussion_r944475635

Should we move all `.expect("DB failure")` instances into our database module, out of application logic?

**Comment by dpc** [comment]
I'm kind of surprised you don't `?` and bubble it up. That's what I would do. Use `color_eyre` for application code, `this-error` for libraries, bubble everything up all the way until it can be either handled or crashes the program and prints a backtrace.

Generally in larger applications with a lot of moving pieces running at the same there's got to be some higher level structure, and methodology to start and shut things down when one of the components failed or the operator wants to shut things down cleanup. At least that's what I always do - bunch of actors (components running in parallel in their own thread(s)), started by the main thread, running at the same time, usually connected via channels and/or some shared state, and some flags indicating that the application wants to shutdown and that one of the components finished unexpectedly. ([Some past blog posts about it](https://dpc.pw/how-i-structure-my-apps-in-rust-and-other-languages#actors), but can also show some code examples in my open source projects).

**Comment by elsirion** [comment]
I agree that ideally we could shut down cleanly, but it hasn't been a big priority since we need to be able to survive hardware failures anyway, so crashing when any invariant is broken seemed like a sensible choice. And the DB having an error is such an invariant break imo, since, to my understanding, for an embedded DB that means the underlying storage medium went missing/was corrupted and there's no way to fix it.

We should eventually implement a cleaner shutdown, which would mean bubbling up any assertion error in a controlled manner, adding a lot of error variants I assume. I'm not sure if it should be a high priority.

For me the question is if in the meantime deduplicating all the `expect("DB error")`s will make the code sufficiently cleaner to do that as a first step instead.

**Comment by dpc** [comment]
> And the DB having an error is such an invariant break imo

Debatable. Embedded KV is an IO, and can fail due to external reasons like running out of disk space. I agree thought, that's it's not a big deal if it panics. 

> For me the question is if in the meantime deduplicating all the expect("DB error")s will make the code sufficiently cleaner to do that as a first step instead.

I don't know the codebase yet, so it's hard for me to judge with confidence the tradeoffs, but bubbling stuff up with `?` to some top level place in every thread/actor is getting rid of `unwrap`/`expect`s and it doesn't require implementing clean shutdown immediately (it will still panic). No strong opinions, but something to consider. Personally, I usually start with `Result`s + `?` right away, because I just find it faster&lazier.

**Comment by m1sterc001guy** [comment]
Resurfacing this discussion. In the code we still crash if the database implementation (rocksdb on the server) encounters any kind of error. When running fedimint we have seen some cases where insert or commit can fail. Some examples:

Commit can fail due to concurrent transactions modifying the same key.
insert_entry can fail (for rocksdb) when writing to the database if the write buffer is out of space (therefore, it cannot determine if there were any write-write conflicts).
Full list of errors that can happen is available here: https://docs.rs/rocksdb/latest/rocksdb/enum.ErrorKind.html

Some of these errors are physical implementation details of rocksdb. Perhaps on the errors where it makes sense to retry, we retry within the rocksdb module, but otherwise we should bubble up the error so that the fedimint-server can handle it?

**Reply by dpc** [reply]
We started using this automatic db commit retry function btw.
